README
NAME

    clad - A simple parallel ssh client.

VERSION

    version 0.26

SYNOPSIS

        clad [log options] [-n] [-a] [-l login] <cluster> <command>

DESCRIPTION

    clad provides the ability to run the same command on several hosts at
    once. The output is displayed unbuffered as the various hosts run the
    command. The list of hosts is determined by reading a configuration
    file which associates names of clusters with lists of hosts. The
    configuration file may also contain command aliases and environment
    settings.

    The command(s) will be executed under '/bin/sh -e' regardless of the
    login shell for the remote user.

FEATURES

      * Cluster-specific and global environment settings

      * Event-driven unbuffered output (using Mojo::Reactor)

      * Host chaining support

      * Perlish configuration for dynamic cluster settings

      * Filtering of banners

      * Command aliases

      * Command macros

OPTIONS

    -n

      Dry run, just show the command that would be executed and each host.

    -a

      Don't colorize the host names in the output.

    -l

      Specify a login name for all ssh connections (proxies, too).

EXAMPLES

     clad -n mailhosts uname -a
     clad webservers df -kh
     clad --trace root webservers ping -i 1 localhost

CONFIGURATION

    The configuration file is a Clustericious::Config file (YAML or JSON
    Mojo::Template's ) and has three sections :

    env : environment settings for all commands run on all hosts.

    clusters : specifies the clusters. A cluster is a label and a list of
    hosts. Each host can be either a hostname or an array of hostnames. If
    it is an array of hostnames it will be treated as a sequence of hosts
    through which to proxy, e.g.

        ssh host1 ssh host2 ssh host3...

    If the first proxy host is the same for all the hosts in a cluster,
    then it can be given separately (see the example below).

    aliases : command aliases. These may be either a single command or a
    list of commands. Lists of commands will all be run in the same ssh
    session. Environment variables in the env section will be merged with
    the global environment settings.

    macros : command macros. These are a sequence of commands or aliases.
    They will be executed in sequence, starting a new connection with each
    one. Each command/alias may also optionally have a login associated
    with it.

    banners : banners which should be suppressed when displaying stderr.

SAMPLE CONFIGURATION

    A sample configuration file ~/etc/Clad.conf :

        ---
        env :
            PATH : /usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/bin
            PERL5LIB : /home/joe/perl5
        clusters :
            mailers :
                - mail1
                - mail2
                - mail3
            webservers :
                env  :
                    USER : bob
                hosts :
                     - web1
                     - web2
            testweb :
                proxy : test.example.com
                hosts :
                    - testweb1
                    - testweb2
        aliases :
            check_ports : lsof -i -n
            startweb : nginx start
            stopweb : nginx start
            restartweb :
                - apachectl stop
                - apachectl start
            config_pull :
                - cd /usr/local/etc
                - git pull
            build :
                - cd project
                - perl Build.PL
                - ./Build
                - ./Build test
                - ./Build install
        macros :
             release :
                - login : joe
                  command : build
                - login : dev
                  command : restartweb
         banners :
                - text : |
    
    
                    **********************************************
                    This is a welcome banner that you see whenever
                    you ssh to our server!
                    **********************************************
                    \n
                    \n

    Another example, using some Mojo::Template features :

        ---
        env :
            PATH : /usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/bin
            PERL5LIB : /home/joe/perl5
        clusters :
            mailers    : [ <%= join ',', map "mail".$_, 0..10 %> ],
            webservers : [ <%= `/usr/local/bin/list_my_web_servers` %> ]

NOTES

    Manipulating .ssh/config can also be used for proxying. For instance :

        Host webdev1
        ProxyCommand ssh firewall.example.com nc %h %p
    
        Host webdev2
        ProxyCommand ssh firewall.example.com nc %h %p

SEE ALSO

      * Clustericious::Admin

      * Clustericious::Config

      * Log::Log4perl::CommandLine

      * SSH::Batch

      * Net::OpenSSH::Parallel

      * Net::SSH::Mechanize

      * helm

AUTHOR

    Original author: Brian Duggan

    Current maintainer: Graham Ollis <plicease@cpan.org>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2012 by NASA GSFC.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Clustericious-Admin

{{$NEXT}}

0.26      2015-11-13 09:39:23EST-0500 America/New_York
  - Fix bug where aliases were not being used

0.25      2014-12-20 05:43:02 America/New_York
  - cpantesters do not report on older Perls

0.24      2014-08-19 16:17:41 America/New_York
  - migrate to Dist::Zilla
  - better documentation coverage
  - do not use ansi color if standard out is not a terminal

0.23 Mon Aug 19 07:36:09 EDT 2013
  - increment version number this time

0.22 Mon Aug 19 07:34:14 EDT 2013
  - minor documentation tweaks

0.21 Mon Nov  5 12:29:52 EST 2012
  - first public release



Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.40
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'Clustericious::Admin',
          'VERSION_FROM' => 'lib/Clustericious/Admin.pm',
          'PREREQ_PM' => {
                           'Clustericious::Config' => 0,
                           'Clustericious::Log' => 0,
                           'Hash::Merge' => 0,
                           'Log::Log4perl::CommandLine' => 0,
                           'Mojolicious' => '2.69',
                           'Term::ANSIColor' => 0,
                           'Test::More' => 0
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => [
                           'bin/clad'
                         ],
          'PL_FILES' => {}
        )
;


lib/Clustericious/Admin.pm
package Clustericious::Admin;

use strict;
use warnings;
use Clustericious::Config;
use Clustericious::Log;
use IPC::Open3 qw/open3/;
use Symbol 'gensym';
use IO::Handle;
use Term::ANSIColor;
use Hash::Merge qw/merge/;
use Mojo::Reactor;
use Data::Dumper;
use Clone qw/clone/;
use v5.10;

# ABSTRACT: Simple parallel ssh client.
# VERSION

=head1 NOTE

B<NOTE>: L<clad> and L<Clustericious::Admin> was rewritten from scratch. 
This repository remains for reference only. The companion module was 
renamed to L<App::clad>.  Please see the new project git repository 
here:

=over 5

=item L<https://github.com/plicease/App-clad>

=back

The remaining documentation is for the original.

=head1 DESCRIPTION

This is a simple parallel ssh client, with a verbose
configuration syntax for running ssh commands on various
clusters of machines.

Most of the documentation is in the command line tool L<clad>.

=cut

our @colors = qw/cyan green/;
our %waiting;
our %filtering;
our $SSHCMD = "ssh -o StrictHostKeyChecking=no -o BatchMode=yes -o PasswordAuthentication=no";

sub _conf {
    our $conf;
    $conf ||= Clustericious::Config->new("Clad");
    return $conf;
}

=head1 FUNCTIONS

=head2 banners

 my $banners = Clustericious::Admin->banners;

return the banners from the configuration file as a list.

=cut

sub banners {
    our $banners;
    $banners ||= _conf->banners(default => []);
    for (@$banners) {
        $_->{text} =~ s/\\n/\n/g;
        my @lines = $_->{text} =~ /^(.*)$/mg;
        $_->{lines} = \@lines;
    }
    return @$banners;
}

sub _is_builtin {
    return $_[0] =~ /^cd /;
}

sub _queue_command {
    my ($user,$w,$color,$env,$host,@command) = @_;
    TRACE "Creating ssh to $host";

    my($wtr, $ssh, $err);
    $err = gensym;
    my $ssh_cmd;
    my $login = $user ? " -l $user " : "";
    if (ref $host eq 'ARRAY') {
        $ssh_cmd = join ' ', map "$SSHCMD $login $_", @$host;
        $host = $host->[1];
    } else {
        $ssh_cmd = "$SSHCMD $login -T $host";
    }
    my $pid = open3($wtr, $ssh, $err, "trap '' HUP; $ssh_cmd /bin/sh -e") or do {
        WARN "Cannot ssh to $host: $!";
        return;
    };

    for my $cmd (@command) {
        my @cmd = $cmd;
        unless (_is_builtin($cmd)) {
            while (my ($k,$v) = each %$env) {
                unshift @cmd, "$k=$v";
            }
            unshift @cmd, qw/env/;
        }
        print $wtr "@cmd\n";
    }

    TRACE "New ssh process to $host, pid $pid";
    $waiting{$host} = $pid;

    $w->io( $ssh,
        sub {
            if (eof($ssh)) {
                TRACE "Done with $host (pid $waiting{$host}), removing handle";
                $w->remove($ssh);
                delete $waiting{$host};
                $w->stop unless keys %waiting > 0;
                return;
            }
            chomp (my $line = <$ssh>);
            print color $color if @colors;
            print "[$host] ";
            print color 'reset' if @colors;
            print "$line\n";
         });

    my $banners = [banners()];
    $w->io(
        $err,
        sub {
            state $filters = [];
            return if eof($err);
            my $skip;
            chomp (my $line = <$err>);
            for (0..$#$banners) {
                $filters->[$_] //= { line => 0 };
                my $l = \( $filters->[$_]{line} );
                my $filter_line = $banners->[$_]{lines}[$$l];
                if ($line eq $filter_line) {
                    TRACE "matched filter $_ (line $$l) : '$line'";
                    $skip = 1;
                    $$l++;
                    if ($$l >= @{ $banners->[$_]{lines} }) {
                        $filters->[$_] = undef;
                    }
                } else {
                    $filters->[$_] &&= undef;
                    TRACE "line vs filter number $_ : '$line' vs '$filter_line'";
                }
            }
            return if $skip;
            print color $color if @colors;
            print "[$host (stderr)] ";
            print color 'reset' if @colors;
            print "$line\n";
         });
    $w->watch($err,1,0);
    $w->watch($ssh,1,0);
    $w->on(error => sub {
        my ($reactor,$err) = @_;
        ERROR "$host : $err";
        delete $waiting{$host};
        $reactor->stop unless keys %waiting > 0;
    });

}

=head2 clusters

 my @clusters = Clustericious::Admin->clusters;

Return the list of clusters in the configuration file

=cut

sub clusters {
    my %clusters = _conf->clusters;
    return sort keys %clusters;
}

=head2 aliases

 my @aliases = Clustericious::Admin->aliases;

Return the list of aliases from the configuration file

=cut

sub aliases {
    my %aliases = _conf->aliases(default => {});
    return sort keys %aliases;
}

=head2 run

 Clustericious::Admin->run(\%options, $cluster, $command);

Run the given command on all the hosts in the given cluster.
Returns the exit value for the equivalent clad command, which
is currently always 0, but in the future might be non-zero
for failure.  Options include:

=over 4

=item n

Set to true for a dry run

=item l

Set to the username that you want to connect to if not using the default.

=item a

Set to true to turn off color

=back

=cut

sub run {
    my(undef, $opts, $cluster, @cmd) = @_;
    my $dry_run = $opts->{n};
    my $user = $opts->{l};
    local @colors = @colors;
    @colors = () if $opts->{a};
    @colors = () unless -t STDOUT;
    LOGDIE "Missing cluster" unless defined $cluster;
    my $clusters = _conf->clusters(default => '') or LOGDIE "no clusters defined";
    ref($clusters) =~ /config/i or LOGDIE "clusters should be a yaml hash";
    my $hosts = $clusters->$cluster(default => '') or LOGDIE "no hosts for cluster $cluster";
    my $cluster_env = {};
    my @hosts;
    if (ref $hosts eq 'ARRAY') {
        @hosts = @$hosts;
    } else {
        @hosts = $hosts->hosts;
        if (my $proxy = $hosts->proxy(default => '')) {
            @hosts = map [ $proxy, $_ ], @hosts;
        }
        $cluster_env = $hosts->{env} || {};
    }
    LOGDIE "no hosts found" unless @hosts;
    my $alias = $cmd[0] or LOGDIE "No command given";

    my @command;
    if (my $command = _conf->aliases(default => {})->{$alias}) {
        DEBUG "Found alias $alias";
        @command = ref $command ? @$command : ( $command );
    } else {
        DEBUG "No alias $alias using @cmd";
        @command = @cmd;
    }
    LOGDIE "No command" unless @command && $command[0];
    s/\$CLUSTER/$cluster/ for @command;
    DEBUG "Running @command on cluster $cluster";
    my $i = 0;
    my $env = _conf->{env} || {};
    $env = merge( $cluster_env, $env );
    TRACE "Env : ".Dumper($env);
    my $watcher = Mojo::Reactor->detect->new;
    for my $host (@hosts) {
        $i++;
        $i = 0 if $i == @colors;
        my $where = ( ref $host eq 'ARRAY' ? $host->[-1] : $host );
        if ($dry_run) {
            INFO "Not running on $where : " . join '; ', @command;
        } else {
            TRACE "Running on $where : " . join ';', @command;
            _queue_command( $user, $watcher, $colors[$i], $env, $host, @command );
        }
        if ( Log::Log4perl::get_logger()->is_trace ) {
            $watcher->recurring(
                2 => sub {
                    TRACE "Waiting for $host (pid $waiting{$host})" if $waiting{$host};
                    TRACE "Not waiting for any host" unless keys %waiting;
                }
            );
        }
    }
    $watcher->recurring(1 => sub { TRACE "tick" } );
    $watcher->start unless $dry_run;
    
    0;
}

1;

=head1 SEE ALSO

L<clad>

=head1 TODO

Handle escaping of quote/meta characters better.

=cut


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.039.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "Simple parallel ssh client.",
  "dist_author" => [
    "Graham Ollis <plicease\@cpan.org>"
  ],
  "dist_name" => "Clustericious-Admin",
  "dist_version" => "0.26",
  "license" => "perl",
  "module_name" => "Clustericious::Admin",
  "recursive_test_files" => 1,
  "requires" => {
    "Clone" => 0,
    "Clustericious::Config" => 0,
    "Clustericious::Log" => 0,
    "Data::Dumper" => 0,
    "Hash::Merge" => 0,
    "IPC::Open3" => 0,
    "Log::Log4perl::CommandLine" => 0,
    "Mojolicious" => "2.69",
    "Symbol" => 0,
    "perl" => "v5.10.0"
  },
  "script_files" => [
    "bin/clad"
  ],
  "test_requires" => {
    "Test::Clustericious::Config" => 0,
    "Test::More" => 0
  }
);


my %fallback_build_requires = (
  "Module::Build" => "0.28",
  "Test::Clustericious::Config" => 0,
  "Test::More" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);


$build->create_build_script;
